src/data/sequence.rs

```diff
@@ -0,0 +1,603 @@
+
+    /// Returns a range in the history of data.
+    pub fn in_range(&self, start: Version, end: Version) -> Option<Values> {
+        state_dispatch!(self, ref state => state.in_range(start, end))
+    }
+
+    /// Returns the owner at a specific version of owners.
+    pub fn owner_at(&self, version: impl Into<Version>) -> Option<&Owner> {
+        state_dispatch!(self, ref state => state.owner_at(version))
+    }
+
+    /// Returns history of all owners
+    pub fn owner_history(&self) -> Result<Vec<Owner>> {
+        state_dispatch!(self, ref state => Some(state.owner_history())).ok_or(Error::NoSuchEntry)
+    }
+
+    /// Get history of owners within the range of versions specified.
+    pub fn owner_history_range(&self, start: Version, end: Version) -> Result<Vec<Owner>> {
+        state_dispatch!(self, ref state => state.owner_history_range(start, end))
+            .ok_or(Error::NoSuchEntry)
+    }
+
+    /// Returns a specific user's access list of a public instance at a specific version.
+    pub fn public_user_access_at(
+        &self,
+        user: User,
+        version: impl Into<Version>,
+    ) -> Result<PublicUserAccess> {
+        self.public_access_list_at(version)?
+            .access_list()
+            .get(&user)
+            .cloned()
+            .ok_or(Error::NoSuchEntry)
+    }
+
+    /// Returns a specific user's access list of a private instance at a specific version.
+    pub fn private_user_access_at(
+        &self,
+        user: PublicKey,
+        version: impl Into<Version>,
+    ) -> Result<PrivateUserAccess> {
+        self.private_access_list_at(version)?
+            .access_list()
+            .get(&user)
+            .cloned()
+            .ok_or(Error::NoSuchEntry)
+    }
+
+    /// Returns the access list of a public instance at a specific version.
+    pub fn public_access_list_at(&self, version: impl Into<Version>) -> Result<&PublicAccessList> {
+        use Sequence::*;
+        let access_list = match self {
+            Public(data) => data.access_list_at(version),
+            _ => return Err(Error::InvalidOperation),
+        };
+        access_list.ok_or(Error::NoSuchEntry)
+    }
+
+    /// Returns the access list of a private instance at a specific version.
+    pub fn private_access_list_at(
+        &self,
+        version: impl Into<Version>,
+    ) -> Result<&PrivateAccessList> {
+        use Sequence::*;
+        let access_list = match self {
+            Private(data) => data.access_list_at(version),
+            _ => return Err(Error::InvalidOperation),
+        };
+        access_list.ok_or(Error::NoSuchEntry)
+    }
+
+    /// Returns history of all access list states
+    pub fn public_access_list_history(&self) -> Result<Vec<PublicAccessList>> {
+        use Sequence::*;
+        let result = match self {
+            Public(data) => Some(data.access_list_history()),
+            _ => return Err(Error::InvalidOperation),
+        };
+        result.ok_or(Error::NoSuchEntry)
```

Comment:
I think `if let` would be a better fit here.
```
if let Sequence::Public(data) = self {
    Ok(data.access_list_history())
} else {
    Err(Error::InvalidOperation)
}
```
Also I'm not sure what kind of error we want to return here. If the type is private it will return `InvalidOperation` but in the above code result is always `Some` so `NoSuchEntry` will never be returned here. Not sure what was intended here.

Reply:
I see a similar pattern in some other functions too.

Reply:
Great observation. I basically tried with first pattern that came to mind on those places. This is naturally much better.
Are there more places to fix now after the latest commits?